### PR TITLE
Give nicer lambda variables

### PR DESCRIPTION
**Pull request by forki**
Since we are in the age of fable, we should care more about generated names.
This is a first attempt to make https://github.com/fable-compiler/Fable/issues/713 nicer

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
As I understand, it just makes generated JS looks nicer? Why should we care as nobody looks at it?

**Comment by forki** [comment]
we actually look at it all the time. You need to debug running apps and often you don't have sourcemaps. Even when we have source maps we want to look at the generated js from time to time.

**Comment by forki** [comment]
@vasily-kirichenko 

> F# |> BABEL - The compiler that emits JavaScript you can be proud of!

that's actually really important for us as Fable users. Not because of being proud, but because this allows us to still reason about the code. 

One interesting thing is that today @alfonsogarciacaro and myself compiled my 97 line F# Sudoku solver sample to just 106 lines of (really nice) JavaScript. This is basically nearly a 1:1 ratio where I'd have expected 1:10 or something like that. I wonder if we find a handwritten JavaScript sample that is shorter and still readable.
BTW: the craziest thing is: we compiled it in the browser! I mean that F# compiler running in the browser story is no fake news. It's really working.

**Comment by vasily-kirichenko** [comment]
:) However I personally don't see any value in "everything works in the browser"...

**Comment by forki** [comment]
But I'm sure you would appreciate the fact that obfuscation and compilation are two distinct things ;-)

**Comment by KevinRansom** [comment]
@forki in what way are they "nicer" after this.  

As a team we are in favour of making sure that transpiling to JS gives a good JS developer experience, and making the right changes in the FSharp compiler to support that.

Before and Afters of names would be useful.  

And perhaps some tests so that we can ensure that names stay nice ?

**Comment by forki** [comment]
I don't know how to write a test, since this is only in generated IL or
generated Javascript (which only see in fable).

The improvement is that in some lambdas we now have argument name "x"
instead of "arg00". This looks a bit less computer generated.

Am 15.02.2017 22:25 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
1.  I'm not sure that "x" is better than "arg00", just different.

2.  You can use ilasm to generate a text file.  Or perhaps use reflection to grab the generated method and ensure the name is one that you expect.

3.  Perhaps there is some way to allow the lambda to be annotated, maybe with the CompiledName attribute so that the developer can specify a meaningful name.

**Comment by forki** [comment]
Please look at https://github.com/fable-compiler/Fable/issues/713 - there is just no place where we can set a compiled name attribute since this is just not existent in the original code. 

Regarding x vs arg00 - I will now just use arg00 for all lambda args in upcoming pull requests. ;-)

**Comment by cartermp** [comment]
@KevinRansom `x` is most assuredly better than `arg00`, especially because the generated JS from FABLE is intended to be read (and perhaps even modified) as "real" JS and not just some intermediate language to run in the browser.  `x` is idiomatic JS, but `arg00` is not.

**Comment by KevinRansom** [comment]
@cartermp they are both identifiers right?  x is an idiomatic JavaScript identifier and arg00 is not an idiomatic javascript identifier,

Can you explain that more fully?

**Comment by cartermp** [comment]
It's not so much `x` so much as it's *not* `arg00`.  `x` is used in multiple examples on the web for JS Arrow Functions as an example of a scoped variable in a closure.  `arg00` is not shown anywhere and looks like it was generated by some tool.  One of the goals of FABLE is to generate JS code that doesn't look like it came out of a tool.  This PR helps that.

**Comment by smoothdeveloper** [comment]
Should we have tests to assert on the expected behaviour?

Should it generate x, y then z, then arg4, arg5 etc.?

**Comment by enricosada** [comment]
@smoothdeveloper let's go excel style.. `A`, `B`, `C`, .. , `Z`, `AA`...`ZZ`, `AAA`...

**Comment by smoothdeveloper** [comment]
If we are going to try to avoid arg00, can we use the type of parameters? if it is a function, f, g, etc.

I think we should have a test to at least describe and pin the behaviour without having to debug in that spot, or at least a comment explaining what names comes out.

File: src/fsharp/TypeChecker.fs [hunk 1]
```diff
@@ -9346,7 +9346,14 @@ and TcMethodApplication
                             [domainTy]
                     [argTys],returnTy
                         
-            let lambdaVarsAndExprs = curriedArgTys |> List.mapiSquared (fun i j ty -> mkCompGenLocal mMethExpr ("arg"+string i+string j) ty)
+            let lambdaVarsAndExprs =
```
Comment (smoothdeveloper) [inline src/fsharp/TypeChecker.fs]: Would you consider commenting the expected naming logic happening here?

Comment (smoothdeveloper) [inline src/fsharp/TypeChecker.fs]: (I mean adding a comment in the source).

**Comment by KevinRansom** [comment]
So the current PR address the non-curried argument scenario or a single argument I suppose.   With curried args to the lambda it is still arg+string I + string j.

What happens if there is a developer supplied identifier of x, I assume that would need discriminating.

perhaps we can provide a #pragma or something that will tell the compiler the base name of lambdas, perhaps also we can auto strip the discriminator for the first occurrence ie. it's zero.

Alternatively the rename could also happen at the point of code generation.

I think we need to think this through, after all it will represent a compiler commitment on naming, and one that we won't easily be able to change if we get it wrong.  So perhaps a spec, scenarios and discussion somewhere.

We are currently fortunate that it doesn't really matter what we call them for IL.

Kevin

**Comment by forki** [comment]
Regarding nameclash. Same thing applies to someone calling things arg00.
Fwiw this is not meant as: we need to do this right now. This is actually
more a proof of concept. There are a couple more "weird" names hidden from
the compiler. I'm planning to address more of these.

Am 16.02.2017 19:27 schrieb "Kevin Ransom (msft)" <notifications@github.com

**Comment by KevinRansom** [comment]
@forki

Yes I know that arg00 will have the same collision.  That doesn’t make it right to collide and is something that will need fixing for transpiling to be a real thing.  I don’t think it is a problem for the F# compiler because in il we don’t really use the names, instead indexes.  Although I suppose the debugger would have a hard time with local variables that collide.

It’s great to have this discussion, I am urging that we consider the whole problem.  It may also be preferable that the renaming happens later when the JavaScript code is emitted.  I don’t know whether that is preferable.

Kevin

**Comment by alfonsogarciacaro** [comment]
I heard about F#, Fable and JS so I feel that I must say something :)

- **Nice names for variables**: I commented in fable-compiler/Fable#713 that the argument names were generated by the F# compiler, but that's a bit of an excuse because Fable can in fact change them (if the `IsCompilerGenerated` property evals to true). Personally I didn't consider the issue that much at the time (in fact, Fable also generates ugly var names in some places) but I understand @forki points in making the code more readable.

- **Name collision**: I had this problem in the very first version of Fable, and I did a PR to append a digit to some variable names. But after talking to @dsyme and as @KevinRansom confirms, I realized this name collision doesn't affect F# IL code, so I introduced reference checks for F# variables and a [custom identifier name resolution](https://github.com/fable-compiler/Fable/blob/4f3b3cccaf088c14bc9a07123bfa4ba1f45704b2/src/dotnet/Fable.Core/Util.fs#L134-L138) for Fable. There shouldn't be conflicts in the generated JS code.

In conclusion, although this PR would make Fable's life (and mine) easier, if it's considered somewhat risky, this can still be dealt with in Fable.

BTW, Fable also has some heuristics on the names generated by the F# compiler: basically it assumes that [compiler generated variables named `tupledArg`](https://github.com/fable-compiler/Fable/blob/4f3b3cccaf088c14bc9a07123bfa4ba1f45704b2/src/dotnet/Fable.Compiler/FSharp2Fable.Util.fs#L490) will immediately be deconstructed, so it's also in Fable interest to be careful with the name generation.

**Comment by KevinRansom** [comment]
I'm going to close this for now, as it seems to not offer sufficient benefit.  And the renaming can occur in fable.
